<a href="https://colab.research.google.com/github/organisciak/Scripting-Course/blob/master/labs/06-visualization-lab.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Lab 6 - Part 1

~This week's lab will be submitted in two weeks, combined with next week's lab.~ *(Part 2 will be ungraded in 2020, though you still have two weeks to submit Part 1)*

*Reminder - save your work. Go to File > Save a Copy in Drive to ensure that you have your work saved.*

As we move toward your final projects, we're also practicing our 'figure out new stuff' skills. Remember the high level tips: look at documentation in Colab, try auto-fill to see what methods exist, search Google, search Stack Overflow, and read the documentation for the library (i.e. at http://pandas.pydata.org/).

In [0]:
import pandas as pd

## Dataset: Life Expectancy By Year, Race, Gender

Load this CDC dataset, _Death rates and life expectancy at birth_. Details about the dataset are at [Data.gov](https://catalog.data.gov/dataset/age-adjusted-death-rates-and-life-expectancy-at-birth-all-races-both-sexes-united-sta-1900).

In [27]:
data = pd.read_csv('https://data.cdc.gov/api/views/w9j2-ggv5/rows.csv?accessType=DOWNLOAD')
# Rename columns
data = data.rename(columns={"Average Life Expectancy (Years)": "LifeExp", "Age-adjusted Death Rate": "DeathRate"})
# Select up to 2017
data = data[data.Year <= 2017]
data.sample(10)

Year       Race         Sex  LifeExp  DeathRate
739  1931      White  Both Sexes     62.6     1832.5
143  1925  All Races      Female     60.6     1957.1
674  1984      Black        Male     65.3     1597.1
923  1997      White      Female     79.9      707.5
518  1946      Black      Female     61.0     1556.8
961  1917      White        Male     49.3     2375.9
333  1997  All Races        Male     73.6     1087.7
346  2010  All Races        Male     76.2      887.1
184  1966  All Races      Female     73.9     1055.3
16   1916  All Races  Both Sexes     51.7     2266.6

**Questions**

![Life Expectancy](https://github.com/organisciak/Scripting-Course/blob/master/images/lifeexpectancy.png?raw=1)

Q1) Which Race+Gender is represented in the above plot (*3pts*):
  - Black / Male
  - White / Both Sexes
  - White / Female
  - Black / Both Sexes

(Clarification: There are two charts that look very similar, but only one of them is an answer option!)

In [0]:
q1_answer = "" #@param ["", "a) Black/Male", "b) White/Both Sexes", "c) White/Female", "d) Black/Both Sexes"]

Q2) Write code to recreate the above plot, but giving the plot a sensible title, like below. This will require a skill not mentioned in class.  (*5pts*)

In [0]:
# Answer-Q2




![](https://github.com/organisciak/Scripting-Course/blob/master/images/lifeexp2.png?raw=1)

Q3) Modify the above code to add a y-axis label, *Life Expectancy (Years)*. You can't change the axis labels within `plot()`. However, you can set the output of a `plot()` method to a variable, and that variable will have a method to *set* the *y-axis*.  (*6pts*)


In [0]:
# Answer-Q3


Q4) Write the code to plot a histogram of the life expectancy values across all conditions and years. (*4pts*)


In [0]:
# Answer-Q4



Q5) Which Race+Sex group had the lowest life expectancy, and in what year?  (*4pts*)
  - Black Male in 1907
  - Black Female in 1918
  - Black Male in 1904

In [0]:
q5_answer = "" #@param ["", "a) Black Male in 1907", "b) Black Male in 1918", "c) Black Male in 1904"]

In [22]:
data.groupby(['Race', 'Sex'])['LifeExp'].mean().round(1)

Race       Sex       
All Races  Both Sexes    66.9
           Female        69.5
           Male          64.3
Black      Both Sexes    58.8
           Female        61.4
           Male          56.2
White      Both Sexes    67.6
           Female        70.3
           Male          65.1
Name: LifeExp, dtype: float64



Q6) To one precision point (e.g. XX.X), fill in the blanks for mean life expectancy throughout the entire span of time:  (*6pts*)

<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th></th>
      <th>LifeExp</th>
    </tr>
    <tr>
      <th>Race</th>
      <th>Sex</th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th rowspan="3" valign="top">All Races</th>
      <th>Both Sexes</th>
        <td><bold>BLANK_A</bold></td>
    </tr>
    <tr>
      <th>Female</th>
      <td>69.5</td>
    </tr>
    <tr>
      <th>Male</th>
      <td>64.3</td>
    </tr>
    <tr>
      <th rowspan="3" valign="top">Black</th>
      <th>Both Sexes</th>
      <td>58.8</td>
    </tr>
    <tr>
      <th>Female</th>
      <td>61.4</td>
    </tr>
    <tr>
      <th>Male</th>
      <td><bold>BLANK_B</bold></td>
    </tr>
    <tr>
      <th rowspan="3" valign="top">White</th>
      <th>Both Sexes</th>
      <td>65.1</td>
    </tr>
    <tr>
      <th>Female</th>
      <td><bold>BLANK_C</bold></td>
    </tr>
    <tr>
      <th>Male</th>
      <td>64.8</td>
    </tr>
  </tbody>
</table>

In [0]:
q6a_answer = 0 #@param {type:"number"} 
q6b_answer = 0 #@param {type:"number"} 
q6c_answer = 0 #@param {type:"number"} 

Q7) In point-form steps, how might you get average life expectancy by decade for All Races / Both Sexes? I'm interested in how you think through the solution - code is optional.  (*7pts*)

In [0]:
q7_answer = '''

Type your answer here (the triple quotes tell Python that this is a 
multi-line string, by the way, so keep those at the start and end)


'''



Q8) Read the Seaborn documentation for FacetGrid (docs at https://seaborn.pydata.org/), and write the code to mimic the following chart. (Note, the lines will look different: I modified the data here so that it doesn't give away an earlier answer).  (*10pts*)
![](https://github.com/organisciak/Scripting-Course/blob/master/images/facetgrid-lineplot.png?raw=1)

In [0]:
# Answer-Q8


# Submission Instructions

In [0]:
#@markdown ### First, Enter your name for grading
my_name = "" #@param { type:'string' }

#@markdown _Have you saved your work for yourself? Don't forget to Save a Copy in Drive so that you have your progress._

In [0]:
#@markdown ### Second, check your work:

#@markdown - have you answered all the questions?
#@markdown - Does this notebook run from top to bottom?
#@markdown     - Go to "Runtime > Restart and run all..." to check. Do all the cells run, to the very bottom, or is there a cell in the middle with an error?
#@markdown - Have you completed all the answers where you entered code, keeping the `# Answer-Qx` line at the start of those cells?

#@markdown *A lab that the professor has to fix manually will lose 10pts - run the checks!*

#@markdown ### Finally, submit it

#@markdown - Download the file with "File > Download .ipynb" and submit it to the Canvas assignment page.